FWCore/Framework/src/ProductResolverBase.cc

```diff
@@ -30,6 +30,18 @@ namespace edm {
 
   Provenance const* ProductResolverBase::provenance() const { return provenance_(); }
 
+  void ProductResolverBase::connectTo(ProductResolverBase const&, EventTransitionInfo const&) { assert(false); }
+  void ProductResolverBase::connectTo(ProductResolverBase const&, LumiTransitionInfo const&) { assert(false); }
+  void ProductResolverBase::connectTo(ProductResolverBase const&, RunTransitionInfo const&) { assert(false); }
+  void ProductResolverBase::connectTo(ProductResolverBase const&, ProcessBlockTransitionInfo const&) { assert(false); }
+
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, EventTransitionInfo const&) const { assert(false); }
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, LumiTransitionInfo const&) const { assert(false); }
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, RunTransitionInfo const&) const { assert(false); }
+  void ProductResolverBase::prefetchAsync_(PrefetchArguments&, ProcessBlockTransitionInfo const&) const {
```

Comment:
These defaults are telling me we are approaching this problem in the wrong way. A given ProductResolver will never be used by more than one kind of Principal. Therefore such a virtual overload set doesn't make any sense. It looks to me like the ProductResolver's should become templated on the Principal type for which they are going to be used.

Reply:
Should the `ProductResolverBase` be also templated on the `Principal` type?

Reply:
>Should the ProductResolverBase be also templated on the Principal type?

Yes, that is what I meant to say.

Reply:
Does this also imply that both of you are saying you agree with the percolation approach? And I should go ahead and start working on making ProductResolverBase a template class. That seems a reasonable approach to me.